qualtran/bloqs/generalized_qsp.py

```diff
@@ -239,9 +260,11 @@ def safe_angle(x):
 class GeneralizedQSP(GateWithRegisters):
```

Comment:
The can be implemented as a frozen dataclass using `@attrs.frozen`

Reply:
It is already marked `@frozen`

Reply:
Sorry, I missed that. We shouldn't need an explicit hash function then?

Reply:
the members `P, Q` could be `list`s or `NDArray`s, which are not hashable. So I had to work-around by hashing the essential data for the bloq (U + signal rotation sub-bloqs)

Reply:
ah yes, that works! fixed now